### PR TITLE
feat(condo): DOMA-9613: Add recipients counter to cross-posting integrations

### PR DESCRIPTION
**Pull request by toplenboren**
# Add recipients counter to cross-posting integrations

image

## Context

Context: we want to add ability to check out number of recipients before you send cross-posted news item. 

## Solution

I've implemented custom query that allows our partners to give their users this data. This is not a final solution, as details and architecture may be changed, (thus it is Proof of work)

## How it works 

- Our partner / Open source contributor / 3rd party wants to share news to his platform of choice using Condo as UI and Backend. They already have integration working (Info on `newsSharing` is outside of the scope of this PR)
- Our partner wants to give their users number of subscribers (recipients) so that users knows how many people will get his message

To do this they implement `POST` method called `customGetRecipientsCounters` (or whatever they want to call it) and put a url link to service that expose this API to `newsSharingConfig`

Example url: `https://my-news-sharing-api/getNewsInfo`

We expect that this method accepts this data:

```
{
  organization: {
      tin: string,
      id: string,
      name: string,
  },
  
  scopes: [{
      property: string,
      unitName: string,
      unitType: string,
  }]

  properties: [{
      id: string,
      address: string,
      addressMeta: addressMeta (JSON),
  }]
}
```

Condo expects output in this format:

```
{
  receiversCount: number
}
```

## Note:

This method is created to support only advanced usecases with custom UI and custom Scopes logic. Integrations that do not require custom UI + custom Scopes logic (such as Telegram / Viber) would not require this logic.

More simple solution will be introduced later this year

# Bonus

Sometimes you do not want `preview` to be shown, for example when your app does not have any preview (maybe internal system?)

Sometimes you do not want to show preview for push messages. For example, if your app sends push messages ONLY for emergency news items

I've introduced a setting in `newsSharingConfig` that covers this usecase

### CODE REVIEW

File: apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+const { checkPermissionsInEmployedOrRelatedOrganizations } = require('@condo/domains/organization/utils/accessSchema')
+
+async function canGetNewsSharingRecipientsCounters ({ authentication: { item: user }, context, args: { data: { b2bAppContext: { id: b2bAppContextId } } } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+    if (user.isAdmin || user.isSupport) return true
+
+    const b2bContext = await getById('B2BAppContext', b2bAppContextId)
+    if (!b2bContext) return false
```
Comment (Alllex202) [inline apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js:17]: Should check `status` and `deletedAt`

File: apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    const organizationId = get(b2bContext, 'organization', null)
+    if (!organizationId) return false
+
+    return await checkPermissionsInEmployedOrRelatedOrganizations(context, user, organizationId, 'canManageNewsItems')
```
Comment (Alllex202) [inline apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js:22]: Maybe `canReadNewsItems`?
If I only have read access to news then i should see this counter too

File: apps/condo/domains/news/components/NewsForm/NewsItemSharingForm.tsx [hunk 3]
```diff
@@ -13,6 +13,9 @@ import { IFrame } from '@condo/domains/miniapp/components/IFrame'
 import { MemoizedSharingNewsPreview } from '@condo/domains/news/components/NewsPreview'
 import { NEWS_TYPE_EMERGENCY } from '@condo/domains/news/constants/newsTypes'
 
+import { NewsItemSharingCustomRecipientCounter } from '../RecipientCounter'
+import { TNewsItemScopeNoInstance } from '../types'
```
Comment (Alllex202) [inline apps/condo/domains/news/components/NewsForm/NewsItemSharingForm.tsx:17]: For names of types and interfaces we usually use prefix `I` or postfix `Type`. 
I have not yet encountered prefix `T`

Comment (toplenboren) [reply apps/condo/domains/news/components/NewsForm/NewsItemSharingForm.tsx:17]: Agreed that this syntax is a bit wierd :-) Sometimes it is used in condo: TTextHighlighterRenderPartFN

Fixed it

File: apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js (hunk 2 above)
Comment (Alllex202) [reply apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js:23]: Ok, I get it
This block should not be displayed there without write permissions

**Comment by Alllex202** [comment]
> This method is created to support only advanced usecases with custom UI and custom Scopes logic. Integrations that do not require custom UI + custom Scopes logic (such as Telegram / Viber) would not require this logic.

For messengers (telegram/viber), you can use the number of group members as the number of recipients

File: apps/condo/domains/miniapp/gql.js [hunk 4]
```diff
@@ -29,7 +29,7 @@ const SEND_B2C_APP_PUSH_MESSAGE_MUTATION = gql`
     }
 `
 
-const B2B_APP_NEWS_SHARING_CONFIG_FIELDS = `{ publishUrl previewUrl customFormUrl getRecipientsUrl icon { publicUrl } previewPicture { publicUrl } name ${COMMON_FIELDS} }`
+const B2B_APP_NEWS_SHARING_CONFIG_FIELDS = `{ publishUrl previewUrl customFormUrl getRecipientsUrl customGetRecipientsCountersUrl icon { publicUrl } previewPicture { publicUrl } name ${COMMON_FIELDS} }`
```
Comment (pahaz) [inline apps/condo/domains/miniapp/gql.js:32]: strange naming. why not just `getRecipientsCunterUrl`?

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 5]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
-            schemaDoc: 'Icon of the app: Telegram Icon / WhatsApp Icon',
+            schemaDoc: 'Icon of the app. For example: Telegram Icon',
             type: 'File',
             isRequired: false,
             adapter: NEWS_SHARING_FILE_ADAPTER,
         },
 
         previewPicture: {
-            schemaDoc: 'Preview picture: might be app screenshot',
+            schemaDoc: 'Preview picture: For example: Telegram screenshot',
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:50]:
```suggestion
            schemaDoc: 'Preview Image: For instance, a screenshot from Telegram',
```

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 6]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
             type: 'File',
             isRequired: false,
             adapter: NEWS_SHARING_FILE_ADAPTER,
         },
 
+        pushNotificationSettings: {
+            schemaDoc: 'Push notification settings',
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:57]:
```suggestion
            schemaDoc: 'Push notification preferences for each B2BApp. The available options include: notifications are sent only for emergency alerts, all notifications are turned off, or all notifications are enabled',
```

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 7]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
+            type: 'Select',
+            options: Object.values(NEWS_SHARING_PUSH_MESSAGE_PREVIEW_SETTINGS),
+            isRequired: true,
+            defaultValue: NEWS_SHARING_PUSH_MESSAGE_PREVIEW_SETTINGS.DISABLED,
+        },
+
+        // Standard set of methods that should be implemented to create an integration with telegram / viber / wa services
+
+        publishUrl: {
+            schemaDoc: 'URL that implements publishing NewsItem method. Check News domain for reference',
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:67]: Write more details for third party developers

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 8]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
+            type: 'Url',
+            isRequired: true,
+        },
+
         previewUrl: {
-            schemaDoc: 'URL that returns HTML preview NewsItem',
+            schemaDoc: 'URL that returns rendered HTML preview of News Item. If not provided, app preview will not be rendered',
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:73]: Where is this URL used? Write some details

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 9]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
+        // Should only be used as a last resort
+
         customFormUrl: {
-            schemaDoc: 'URL that implements customForm. If not filled, then app will use standard news form',
+            schemaDoc: 'URL that implements customForm. Should not be used in regular cases. Allows to provide custom UI for sending news. If not provided app will use condo news form',
+            type: 'Url',
+            isRequired: false,
+        },
+
+        customGetRecipientsCountersUrl: {
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:93]: getRecipientsCountersUrl

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.test.js [hunk 10]
```diff
@@ -32,6 +32,8 @@ describe('B2BAppNewsSharingConfig', () => {
                 publishUrl: faker.internet.url(),
                 previewUrl: faker.internet.url(),
                 getRecipientsUrl: faker.internet.url(),
+                customGetRecipientsCountersUrl: faker.internet.url(),
```
Comment (pahaz) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.test.js:35]:
```suggestion
                getRecipientsCountersUrl: faker.internet.url(),
```

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 11]
```diff
@@ -36,7 +41,24 @@ interface CounterProps {
 
 const styleQuestionCircle: CSSProperties = { color: colors.gray['5'], cursor: 'help' }
 
-const Counter: React.FC<CounterProps> = ({ label, value, type = 'success', hint, downloadButton, isLoading }) => (
+export const RecipientCounterContainer: React.FC<{ title: React.ReactNode }> = ({ children, title }) => {
+    return (
+        <Card>
+            <Typography.Text>
+                { title }
```
Comment (pahaz) [inline apps/condo/domains/news/components/RecipientCounter.tsx:48]: we don't use extra spaces inside this file and others tsx files. Why do you want to add it here?

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 12]
```diff
@@ -36,7 +41,24 @@ interface CounterProps {
...
+            </Typography.Text>
+            <Space direction='vertical' size={24} width='100%'>
+                <Col xs={24}>
+                    <Row align='top' justify='space-evenly'>
+                        { children }
```
Comment (pahaz) [inline apps/condo/domains/news/components/RecipientCounter.tsx:53]: same?

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 13]
```diff
@@ -243,44 +269,94 @@ export const RecipientCounter: React.FC<RecipientCounterProps> = ({ newsItemScop
...
+    const intl = useIntl()
+    const MailingMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.mailing' })
+    const WillReceiveLabelMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.willReceive.label' })
+    const ErrorLoadingMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.error.loading' })
+
+    const [getCounter, { loading: isCounterLoading }] = useLazyQuery(
+        GET_NEWS_SHARING_RECIPIENTS_COUNTERS_QUERY,
+        {
+            onCompleted: (data) => {
+                console.log(data)
```
Comment (pahaz) [inline apps/condo/domains/news/components/RecipientCounter.tsx:318]: debug?

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 14]
```diff
@@ -243,44 +269,94 @@ export const RecipientCounter: React.FC<RecipientCounterProps> = ({ newsItemScop
...
+                console.log(get(data, ['result', 'receiversCount']))
+                setCounter(get(data, ['result', 'receiversCount']))
+            },
+            onError: (error) => {
+                console.error({ msg: 'Failed to load recipients counters', error })
+                const message = get(error, ['graphQLErrors', 0, 'extensions', 'messageForUser'], ErrorLoadingMessage)
```
Comment (pahaz) [inline apps/condo/domains/news/components/RecipientCounter.tsx:324]:
```suggestion
                const message = get(error, ['graphQLErrors', 0, 'extensions', 'messageForUser'], ErrorLoadingMessage) || SOME_DEFAULT_ERROR_TITLE
```

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 15]
```diff
@@ -243,44 +269,94 @@ export const RecipientCounter: React.FC<RecipientCounterProps> = ({ newsItemScop
...
+            },
+        })
+    }, [getCounter, contextId, newsItemScopes])
+
+    // if typeof counter !== number is used here instead of just if !counter because bool(0) => false
+    const isLoadingCounter = isCounterLoading || typeof counter !== 'number'
+
+    return (
+        <RecipientCounterContainer
+            title={<>{MailingMessage}<Typography.Text strong>{buildMessageFromNewsItemScopes(newsItemScopes, intl)}</Typography.Text></>}
```
Comment (pahaz) [inline apps/condo/domains/news/components/RecipientCounter.tsx:355]: it looks hacky. is it possible to move inise `RecipientCounterContainer` or create a special component?

File: apps/condo/domains/news/components/NewsForm/NewsItemSharingForm.tsx [hunk 16]
```diff
@@ -13,6 +13,9 @@ import { IFrame } from '@condo/domains/miniapp/components/IFrame'
 import { MemoizedSharingNewsPreview } from '@condo/domains/news/components/NewsPreview'
 import { NEWS_TYPE_EMERGENCY } from '@condo/domains/news/constants/newsTypes'
 
+import { NewsItemSharingCustomRecipientCounter } from '../RecipientCounter'
+import { NewsItemScopeNoInstanceType } from '../types'
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/news/components/NewsForm/NewsItemSharingForm.tsx:17]: use absolute imports

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 17]
```diff
@@ -20,10 +20,15 @@ import { colors } from '@open-condo/ui/dist/colors'
...
+import {
+    GET_NEWS_ITEMS_RECIPIENTS_COUNTERS_MUTATION,
+    GET_NEWS_SHARING_RECIPIENTS_COUNTERS_QUERY,
+} from '@condo/domains/news/gql'
 import { useNewsItemRecipientsExportToExcelTask } from '@condo/domains/news/hooks/useNewsItemRecipientsExportToExcelTask'
 
-import { TNewsItemScopeNoInstance, TUnit } from './types'
+import { NewsItemScopeNoInstanceType, TUnit } from './types'
+
+import { Loader } from '../../common/components/Loader'
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/news/components/RecipientCounter.tsx:31]: same

File: apps/condo/domains/news/components/RecipientCounter.tsx [hunk 18]
```diff
@@ -243,44 +269,94 @@ export const RecipientCounter: React.FC<RecipientCounterProps> = ({ newsItemScop
...
+    const MailingMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.mailing' })
+    const WillReceiveLabelMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.willReceive.label' })
+    const ErrorLoadingMessage = intl.formatMessage({ id: 'news.component.RecipientCounter.error.loading' })
+
+    const [getCounter, { loading: isCounterLoading }] = useLazyQuery(
+        GET_NEWS_SHARING_RECIPIENTS_COUNTERS_QUERY,
+        {
+            onCompleted: (data) => {
+                console.log(data)
+                console.log(get(data, ['result', 'receiversCount']))
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/news/components/RecipientCounter.tsx:319]: remove console.log

File: apps/condo/migrations/20240729010950-0413_b2bappnewssharingconfig_customgetrecipientscountersurl_and_more.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
...YXBwbmV3c3NoYXJpbmdjb25maWcnLAogICAgICAgICAgICBuYW1lPSdwdXNoTm90aWZpY2F0aW9uU2V0dGluZ3MnLAogICAgICAgICAgICBmaWVsZD1tb2RlbHMuQ2hhckZpZWxkKGNob2ljZXM9WygnT05MWV9FTUVSR0VOQ1knLCAnT05MWV9FTUVSR0VOQ1knKSwgKCdFTkFCTEVEJywgJ0VOQUJMRUQnKSwgKCdESVNBQkxFRCcsICdESVNBQkxFRCcpXSwgZGVmYXVsdD0nRElTQUJMRUQnLCBtYXhfbGVuZ3RoPTUwKSwKICAgICAgICAgICAgcHJlc2VydmVfZGVmYXVsdD1GYWxzZSwKICAgICAgICApLAogICAgICAgIG1pZ3JhdGlvbnMuQWRkRmllbGQoCiAgICAgICAgICAgIG1vZGVsX25hbWU9J2IyYmFwcG5ld3NzaGFyaW5nY29uZmlnaGlzdG9yeXJlY29yZCcsCiAgICAgICAgICAgIG5hbWU9J2N1c3RvbUdldFJlY2lwaWVudHNDb3VudGVyc1VybCcsCiAgICAgICAgICAgIGZpZWxkPW1vZGVscy5UZXh0RmllbGQoYmxhbms9VHJ1ZSwgbnVsbD1UcnVlKSwKICAgICAgICApLAogICAgICAgIG1pZ3JhdGlvbnMuQWRkRmllbGQoCiAgICAgICAgICAgIG1vZGVsX25hbWU9J2IyYmFwcG5ld3NzaGFyaW5nY29uZmlnaGlzdG9yeXJlY29yZCcsCiAgICAgICAgICAgIG5hbWU9J3B1c2hOb3RpZmljYXRpb25TZXR0aW5ncycsCiAgICAgICAgICAgIGZpZWxkPW1vZGVscy5UZXh0RmllbGQoYmxhbms9VHJ1ZSwgbnVsbD1UcnVlKSwKICAgICAgICApLAogICAgICAgIG1pZ3JhdGlvbnMuQWx0ZXJGaWVsZCgKICAgICAgICAgICAgbW9kZWxfbmFtZT0nYjJiYXBwbmV3c3NoYXJpbmdjb25maWcnLAogICAgICAgICAgICBuYW1lPSdnZXRSZWNpcGllbnRzVXJsJywKICAgICAgICAgICAgZmllbGQ9bW9kZWxzLlRleHRGaWVsZChibGFuaz1UcnVlLCBudWxsPVRydWUpLAogICAgICAgICksCiAgICBdCg==
```
Comment (nomerdvadcatpyat) [inline apps/condo/migrations/20240729010950-0413_b2bappnewssharingconfig_customgetrecipientscountersurl_and_more.js:2]: you need to rebase and regenerate migration to update it's number

File: apps/condo/migrations/20240729173935-0414_b2bappnewssharingconfig_customgetrecipientscountersurl_and_more.js [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+exports.up = async (knex) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/migrations/20240729173935-0414_b2bappnewssharingconfig_customgetrecipientscountersurl_and_more.js:4]: Same migration as above. You need to remove it

**Review by nomerdvadcatpyat — APPROVED** [review]
You need to remove duplicated migration and update migration number

File: apps/condo/domains/miniapp/gql.js (hunk 4 above)
Comment (toplenboren) [reply apps/condo/domains/miniapp/gql.js]: Done

File: apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js [hunk 21]
```diff
@@ -39,40 +39,59 @@ const B2BAppNewsSharingConfig = new GQLListSchema('B2BAppNewsSharingConfig', {
...
+
         customFormUrl: {
-            schemaDoc: 'URL that implements customForm. If not filled, then app will use standard news form',
+            schemaDoc: 'URL that implements customForm. Use only if you need custom NewsItemSharing data structure, for example if . Allows to provide custom UI for sending news. If not provided app will use condo news form',
+            type: 'Url',
+            isRequired: false,
+        },
+
+        getRecipientsCountersUrl: {
+            schemaDoc: 'URL that returns number of subscribers for condo scopes. This function should not be used in regular cases. Use only if in your miniapp  Should implement POST customGetRecipientsCounters method. Should not be used in regular cases. Allows to provide custom values for recipients counter. If not provided app will use data from getRecipients. If getRecipients is not provided, app wont render recipients counter',
```
Comment (toplenboren) [inline apps/condo/domains/miniapp/schema/B2BAppNewsSharingConfig.js:94]:
```suggestion
            schemaDoc: 'URL that returns number of subscribers for condo scopes. Should implement POST customGetRecipientsCounters method. Allows to provide custom values for recipients counter. If not provided app will use data from getRecipients. If getRecipients is not provided, recipients counter will not be rendered',
```

File: apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js (hunk 1 above)
Comment (toplenboren) [reply apps/condo/domains/news/access/GetNewsSharingRecipientsCountersService.js]: Done that

**Pull request merged**